File: gameoflife/src/gameoflife.cpp
#include "gameoflife/gameoflife.hpp"

#include <functional>

#include "gameoflife/config.hpp"
#include "utilscpp/matrix.hpp"

GameOfLife::GameOfLife(int rows, int cols,
                       std::function<bool(int, int)> generator)
    : matrix(rows, cols) {
  rule = {2, 3, 3, 3};
  neighborhood_diffs = {
      {-1, -1}, {-1, 0}, {-1, 1}, {0, -1}, {0, 1}, {1, -1}, {1, 0}, {1, 1},
  };
  for (int i = 0; i < rows; i++) {
    for (int j = 0; j < cols; j++) {
      matrix.set(i, j, generator(i, j));
    }
  }
}

GameOfLife::GameOfLife(): matrix(0, 0) {
  rule = {};
  neighborhood_diffs = {};
}

void GameOfLife::update() {
  const Matrix prev_matrix = matrix;
  for (int i = 0; i < matrix.rows; i++) {
    for (int j = 0; j < matrix.cols; j++) {
      int cnt = get_neighborhood_count(i, j, prev_matrix);
      bool is_alive = prev_matrix.get(i, j);
      bool new_state = false;
      if (is_alive && rule.survival_min <= cnt && cnt <= rule.survival_max) {
        new_state = true;
      } else if (!is_alive && rule.birth_min <= cnt && cnt <= rule.birth_max) {
        new_state = true;
      }
      matrix.set(i, j, new_state);
    }
  }
}

int GameOfLife::get_neighborhood_count(int i, int j,
                                       const Matrix &matrix) const {
  int ans = 0;
  for (const auto &diff : neighborhood_diffs) {
    int y = i + diff.first;
    int x = j + diff.second;
    if (Config::SHOULD_MODULE_INDEXES) {
      x = (x + matrix.cols) % matrix.cols;
      y = (y + matrix.rows) % matrix.rows;
    }
    if (!matrix.are_valid_coords(y, x)) continue;
    ans += matrix.get(y, x);
  }
  return ans;
}

void GameOfLife::on_click(int i, int j) { matrix.toggle(i, j); }

File: gameoflifeServer/gameoflife/gameoflife.go
package gameoflife

import (
	"math/rand"
	"time"

	"github.com/DanielMontesGuerrero/simulations/utilsgo"
)

type Rule struct {
	survivalMin int
	survivalMax int
	birthMin    int
	birthMax    int
}

type GameOfLife struct {
	matrix            utilsgo.Matrix
	rule              Rule
	neighborhoodDiffs [][2]int
}

func NewGame(rows int, cols int) *GameOfLife {
	game := new(GameOfLife)
	game.rule.survivalMin = 2
	game.rule.survivalMax = 3
	game.rule.birthMin = 3
	game.rule.birthMax = 3
	game.neighborhoodDiffs = [][2]int{
		{-1, -1},
		{-1, 0},
		{-1, 1},
		{0, -1},
		{0, 1},
		{1, -1},
		{1, 0},
		{1, 1},
	}
	game.matrix = *utilsgo.New(rows+2, cols+2)
	rand.Seed(time.Now().UnixNano())
	for i := 1; i <= rows; i++ {
		for j := 1; j <= cols; j++ {
			if rand.Intn(2) == 1 {
				game.matrix.Set(i, j, true)
			}
		}
	}
	return game
}

func (game *GameOfLife) Update() {
	prevMatrix := game.matrix.Copy()
	for i := 1; i < game.matrix.Rows-1; i++ {
		for j := 1; j < game.matrix.Cols-1; j++ {
			cnt := game.GetNeighborhoodCount(i, j, prevMatrix)
			isAlive := prevMatrix.Get(i, j) == 1
			newState := false
			if isAlive && game.rule.survivalMin <= cnt && cnt <= game.rule.survivalMax {
				newState = true
			} else if !isAlive && game.rule.birthMin <= cnt && cnt <= game.rule.birthMax {
				newState = true
			}
			game.matrix.Set(i, j, newState)
		}
	}
}

func (game *GameOfLife) GetNeighborhoodCount(i int, j int, matrix utilsgo.Matrix) int {
	ans := 0
	for _, diff := range game.neighborhoodDiffs {
		y := i + diff[0]
		x := j + diff[1]
		if !matrix.AreValidCoords(y, x) {
			continue
		}
		ans += matrix.Get(y, x)
	}
	return ans
}

func (game *GameOfLife) Get(i int, j int) int {
	return game.matrix.Get(i, j)
}

func (game *GameOfLife) Println() {
	game.matrix.Println()
}

func (game *GameOfLife) SetBorders(top, bottom, left, right utilsgo.Vector) {
	for i := 0; i < game.matrix.Rows; i++ {
		game.matrix.Set(i, 0, left.Get(i) == 1)
		game.matrix.Set(i, game.matrix.Rows-1, right.Get(i) == 1)
	}
	for i := 0; i < game.matrix.Cols; i++ {
		game.matrix.Set(0, i, top.Get(i) == 1)
		game.matrix.Set(game.matrix.Cols-1, i, bottom.Get(i) == 1)
	}
}

func (game *GameOfLife) ToggleCell(i, j int) {
	game.matrix.Toggle(i, j)
}

func (game *GameOfLife) GetSubgrid(ui, di, lj, rj int) utilsgo.Matrix {
	return *game.matrix.SubMatrix(ui, di, lj, rj)
}
